refactor: reduce DB updates by traits update requests without real changes in the data

devapro opened the pull request:
Thanks for submitting a PR! Please check the boxes below:

- [ ] I have added information to `docs/` if required so people know about the feature!
- [x] I have filled in the "Changes" section below?
- [ ] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

It is common case for mobile app to set traits on the app start, because some properties can be changed offline or by request from another device with the same account etc. However in this case db is update even if no changes in the data. This change should reduce updates by getting list of traits from db than compares the existing and incoming data, and only add/update when the data actually changes. This strategy drastically reduces unnecessary UPDATE operations in case when the clients send requests with the same data.


## How did you test this code?

manually

_Please describe._

zachaysan reviewed (dismissed):
Looks good. Though it seems like CI has failed for this test, perhaps try re-running the test suite with an empty commit?

matthewelwell commented on api/environments/identities/traits/views.py:
```diff
@@ -245,39 +245,65 @@ def bulk_create(self, request):
-                )
 
-            return Response(serializer.data, status=200)
+                return Response(serializer.data if traits else [], status=200)
+
+            return Response(traits, status=200)
```
Hmm... this actually looks like it changes the behaviour of the endpoint which could have an effect on clients. Previously, we are always returning the traits that were sent (assuming they were valid). Now we're sending an empty list if none of the traits are actually updated / created. 

Perhaps we could do something like the following: 

```python
all_traits = [*traits, *SDKCreateUpdateTraitSerializer(instance=existing_traits, many=True).data]
return Response(all_traits)
```

devapro replied:
Thank you for review.
I made changes, now traits always added to the response.
Moved all logic of updating traits to new method, I think it will be a bit easy to read an understand in this case.

matthewelwell reviewed (commented):
> Though it seems like CI has failed for this test, perhaps try re-running the test suite with an empty commit?

This is actually an issue in our CI workflows that don't work correctly for external contributors. We are working to resolve this. 

I've added a comment to a specific part of the diff but, in general, I'd also like to see a test / some tests written for this logic.

matthewelwell commented on api/environments/identities/traits/views.py:
```diff
@@ -273,11 +286,37 @@ def bulk_create(self, request):
+                identity__environment=request.environment,
+            )
+
+            return Response(
+                [
+                    *SDKCreateUpdateTraitSerializer(
+                        instance=all_traits, many=True
+                    ).data,
+                ],
```
```suggestion
            return Response(
                SDKCreateUpdateTraitSerializer(
                    instance=all_traits, many=True
                ).data,
```

matthewelwell reviewed (commented):
Thanks for the changes @devapro . I've added a couple more comments. 

As per my previous review though, it's still missing any sort of testing coverage.

matthewelwell commented:
@devapro just checking in to see if you had time to look at my recent review comments and whether you might have time to update the PR?

devapro commented:
Sorry for long response, will check your comments next week.

devapro commented on api/environments/identities/traits/views.py:
```diff
@@ -273,11 +286,37 @@ def bulk_create(self, request):
+                raise BadRequest("Unable to set traits with client key.")
+
+            # endpoint allows users to delete existing traits by sending null values
+            # for the trait value so we need to filter those out here
```
Moved comment closer to related code.

devapro commented:
> Thanks for the changes @devapro . I've added a couple more comments.
> 
> As per my previous review though, it's still missing any sort of testing coverage.

I checked existing unit tests and looks like it is covered all logic that I refactored.
I think will be good to add test that checks case when db is not called, but my knowledge in python very limited, I didn't find an example of how to do it in current code.

matthewelwell commented:
>I think will be good to add test that checks case when db is not called, but my knowledge in python very limited, I didn't find an example of how to do it in current code.

You can just copy one of the existing tests, manipulate the data to be what you want and then use the `django_assert_num_queries` fixture to do something like: 

```python
with django_assert_num_queries(0):
    ... 
```

You'll find plenty of usages of this fixture throughout the code base.

devapro commented on api/environments/identities/traits/views.py:
```diff
@@ -238,32 +238,47 @@ def increment_value(self, request):
-                raise BadRequest("Unable to set traits with client key.")
+    def _update_traits(self, request):
+
+        identities = {trait["identity"]["identifier"] for trait in request.data}
```
fixed, thanks

devapro commented on api/environments/identities/traits/views.py:
```diff
@@ -238,32 +238,47 @@ def increment_value(self, request):
-        try:
-            if not request.environment.trait_persistence_allowed(request):
-                raise BadRequest("Unable to set traits with client key.")
+    def _update_traits(self, request):
```
:heavy_check_mark:

devapro commented:
> > I think will be good to add test that checks case when db is not called, but my knowledge in python very limited, I didn't find an example of how to do it in current code.
> 
> You can just copy one of the existing tests, manipulate the data to be what you want and then use the `django_assert_num_queries` fixture to do something like:
> 
> ```python
> with django_assert_num_queries(0):
>     ... 
> ```
> 
> You'll find plenty of usages of this fixture throughout the code base.

I can't find existing test for method that I changed. It is not simple for me to understand how to write new test for specific method.

Didn't manage to run tests locally, for running project I am using docker + mounting specific files with changes, because can't install all required dependencies.
